### PR TITLE
MLPAB-2011 - only publish lambda when the lambda source changes

### PR DESCRIPTION
**Pull request by andrewpearce-digital**
# Purpose

Lambda's would previously always publish a new version even when nothing had changed about them.

Fixes MLPAB-2011

## Approach

- Use a terraform_data resource to hold a replacement trigger
- compare SHA256 to replacement trigger to determine when to publish

## Learning

- https://developer.hashicorp.com/terraform/language/resources/terraform-data

### CODE REVIEW

File: terraform/environment/region/modules/s3_antivirus/lambda_function.tf [hunk 1]
```diff
@@ -18,7 +22,7 @@ resource "aws_lambda_function" "lambda_function" {
   runtime          = "provided.al2023"
   timeout          = 300
   memory_size      = 4096
-  publish          = true
+  publish          = filebase64sha256("${path.module}/myFunction.zip") != terraform_data.replacement.triggers_replace[0] ? true : false
```
Comment (hawx) [inline terraform/environment/region/modules/s3_antivirus/lambda_function.tf:25]: wouldn't `!=` give you a `true` or `false`?

**Review by hawx — DISMISSED** [review]

File: terraform/environment/region/modules/s3_antivirus/lambda_function.tf (hunk 1 above)
Comment (andrewpearce-digital) [reply terraform/environment/region/modules/s3_antivirus/lambda_function.tf:25]: If there's a difference between the current sha and what's stored in the trigger, publish = true, is what' I'm going for.

Comment (andrewpearce-digital) [reply terraform/environment/region/modules/s3_antivirus/lambda_function.tf:25]: oh I see what you mean, yes, I could leave it as `a != b`

**Review by acsauk — APPROVED** [review]
:shipit:

**Pull request merged**